### DISCUSSION TITLE
What's the best way to store JWT in NextJS ?

### DISCUSSION THREAD

**Discussion by gjpsiqueira** [vercel/next.js#12610] [Help]
Hi,

I'm developing an app using NextJS. In my page main, I need authenticate by Jwt in API Hotmart to search the courses and show in my page. I am looking a way secure to store JWT in NextJS, because, to use localStorage or Cookie, I can suffer attack.

Any idea?

**Comment by efleurine** [comment]
Hi,

The problem is not spefically related to NextJS I would say.

may be you can use HTTP cookie only, or store your cookie in the database and do the search server side only.

hopefully that will help you.

**Comment by gjpsiqueira** [comment]
Hi,
I need authentication by JWT in another REST API external. So, I having than authentication by API Route NextJS. When I authenticate in route x, how I do can share my token to all routes API NextJS ?

**Comment by leerob** [comment]
Could you provide a little more information about your application? Ideally an example. 

If you communicate server-to-server using Next's API routes, you could have a "service account" JWT stored as an environment variable that would allow you to communicate with external APIs. This is the approach that Google APIs take, for example. 

Here's a [real example of Google Analytics](https://jamstackfns.com/f/google-analytics) so you can see what I mean. 

```js
import { google } from 'googleapis';

export default async (req, res) => {
  const startDate = req.query.startDate;
  const auth = new google.auth.GoogleAuth({
    credentials: {
      client_email: process.env.GOOGLE_CLIENT_EMAIL,
      client_id: process.env.GOOGLE_CLIENT_ID,
      private_key: process.env.GOOGLE_PRIVATE_KEY
    },
    scopes: ['https://www.googleapis.com/auth/analytics.readonly']
  });

  const analytics = google.analytics({
    auth,
    version: 'v3'
  });

  const response = await analytics.data.ga.get({
    'end-date': 'today',
    ids: 'ga:187331021',
    metrics: 'ga:pageviews',
    'start-date': startDate
  });

  return res.status(200).json({
    pageViews: response.data.rows[0][0]
  });
};
```

**Comment by gjpsiqueira** [comment]
Hi, leerob! Thank you for response!
My API NextJS need authentication API external server-to-server by JWT. The problem that token expires in given moment. So, I need to know if token expires and generate new token. How I do store this token to use in another routes API ?

**Reply by leerob** [reply]
That likely means your authentication strategy does not support server-to-server with a private key and instead uses something like OAuth with a refresh token. 

What API specifically are you trying to use?

**Reply by darrylmorley** [reply]
> That likely means your authentication strategy does not support server-to-server with a private key and instead uses something like OAuth with a refresh token. 
> 
> What API specifically are you trying to use?

@leerob What is the best strategy for storing a JWT when using Oauth with a refresh token? Can you write them to a local file when using ISR on Vercel? I'm guessing they won't persist in a memory cache in this scenario (ISR & Vercel)?

**Reply by leerob** [reply]
No, you can't write to a local file in serverless environments. You would need to use a httpOnly cookie.
